casr/src/bin/casr-ubsan.rs

```diff
@@ -0,0 +1,445 @@
+        .num_threads(num_of_threads)
+        .build()
+        .unwrap();
+
+    // Set ubsan env options
+    debug!("Set environment");
+    if let Ok(mut ubsan_options) = env::var("UBSAN_OPTIONS") {
+        if ubsan_options.starts_with(',') {
+            ubsan_options.remove(0);
+        }
+        if ubsan_options.contains("print_stacktrace=0") {
+            ubsan_options = ubsan_options.replace("print_stacktrace=0", "print_stacktrace=1");
+        } else {
+            ubsan_options.push_str(",print_stacktrace=1");
+        }
+        if ubsan_options.contains("report_error_type=0") {
+            ubsan_options = ubsan_options.replace("report_error_type=0", "report_error_type=1");
+        } else {
+            ubsan_options.push_str(",report_error_type=1");
+        }
+        env::set_var("UBSAN_OPTIONS", ubsan_options);
+    } else {
+        env::set_var("UBSAN_OPTIONS", "print_stacktrace=1,report_error_type=1");
+    }
+
+    // Extract ubsan warnings
+    info!("Extracting UBSAN warnings...");
+    info!("Using {} threads", num_of_threads);
+    let warnings: Vec<(&PathBuf, Vec<(UbsanWarning, CrashLine)>)> = custom_pool.install(|| {
+        inputs
+            .par_iter()
+            .filter_map(|input| {
+                let Ok(input_warnings) = extract_warnings(input, &argv, timeout) else {
+                    warn!("Can't to run program with input file {:?}", input);
+                    return None
+                };
+                Some((input, input_warnings))
+            })
+            .collect()
+    });
+
+    info!(
+        "Number of UBSAN warnings: {}",
+        warnings
+            .iter()
+            .map(|(_, input_warnings)| input_warnings.len())
+            .sum::<usize>()
+    );
+
+    // Create report with equal parts for all reports
+    let mut pre_report = CrashReport::new();
+    pre_report.executable_path = argv[0].to_string();
+    let _ = pre_report.add_os_info();
+    let _ = pre_report.add_proc_environ();
+
+    info!("Deduplicating CASR reports...");
+    // Init dedup crashline list
+    let mut crashlines: HashSet<String> = HashSet::new();
+    let mut to_gen: Vec<(PathBuf, UbsanWarning, CrashLine)> = vec![];
+    // Dedup warnings by crashline
+    // Do without paralleling to preserve the specified order
+    let re = Regex::new(r#"(.+:\d+):\d"#).unwrap();
+    for (input, input_warnings) in warnings {
+        for (warning, crashline) in input_warnings {
+            // Drop column number
+            let mut line = crashline.to_string();
+            if let Some(cap) = re.captures(&line) {
+                line = cap.get(1).unwrap().as_str().to_string();
+            }
+            if crashlines.insert(line) {
+                to_gen.push((input.clone(), warning, crashline));
+            }
+        }
+    }
+
+    info!(
+        "Number of UBSAN warnings after deduplication: {}",
+        crashlines.len()
+    );
+
+    // Generate CASR reports
```

Comment:
We should also copy input seeds next to corresponding CASR reports

Reply:
Done